blocklib/limesdr/limesdr/limesdr_cpu.h

```diff
@@ -0,0 +1,588 @@
+                std::cout << fmt::format("{:16b} {:16b}\n",
+                                         stream->read_buffer[j],
+                                         stream->read_buffer[j + 1]);
+            }
+#endif
+        }
+
+        std::vector<std::size_t> triggers;
+        if (has_timing_triggers()) {
+            triggers = find_triggers();
+            // we have more triggers than messages, wait for messages
+            if (triggers.size() > d_timing_messages.size()) {
+                d_pending_triggers = triggers;
+                return work_return_t::OK;
+            }
+        }
+
+        const auto result = flush_samples(wio, produced, triggers);
+        assert(result && *result == available); // must succeed
+        produced += *result;
+        wio.produce_each(produced);
+        return work_return_t::OK;
+    }
+
+    void handle_msg_timing(pmtv::pmt msg) override
+    {
+        const auto map = std::get<std::map<std::string, pmtv::pmt>>(msg);
+
+        try {
+            if (!has_timing_triggers()) {
+                d_timing_messages.clear();
+            }
+
+            d_timing_messages.push_back(
+                { .name = std::get<std::string>(map.at(tag::TRIGGER_NAME)),
+                  .timestamp =
+                      detail::convert_to_ns(std::get<int64_t>(map.at(tag::TRIGGER_TIME))),
+                  .offset = detail::convert_to_ns(
+                      std::get<double>(map.at(tag::TRIGGER_OFFSET)) * 1000000000) });
+
+            const auto last = d_timing_messages.back();
+            d_logger->debug(
+                "Received timing message: name='{}', timestamp={}, offset={}, "
+                "Already Queued={}",
+                last.name,
+                last.timestamp,
+                last.offset,
+                d_timing_messages.size() - 1);
+        } catch (const std::out_of_range& e) {
+            d_logger->error("Could not decode timing message: {}", e.what());
+        }
+    }
+
+private:
+    std::optional<std::size_t> flush_samples(
+        work_io& wio, std::size_t offset, const std::vector<std::size_t>& trigger_offsets)
+    {
+        assert(d_device);
+
+        const auto noutput_items = wio.outputs()[0].n_items - offset;
+
+        std::optional<std::size_t> produced;
+
+        for (std::size_t channel_idx = 0; channel_idx < d_device->streams.size();
+             ++channel_idx) {
+            auto& stream = d_device->streams[channel_idx];
+            if (!stream) {
+                continue;
+            }
+
+            if (noutput_items < stream->samples_read) {
+                return std::nullopt;
+            }
+
+            auto out = wio.outputs()[channel_idx * 2].items<int16_t>();
+            assert(!produced || *produced == stream->samples_read);
+            produced = stream->samples_read;
+            memcpy(out + offset * 2,
+                   stream->read_buffer.data(),
+                   *produced * 2 * sizeof(int16_t));
```

Comment:
Here we just copy the data received by LMS_RecvStream unmodified. Is this the format we want to output, or should we convert it to something else?

Reply:
There should be a 2 stage process with either the detection of the trigger closer to the driver or closer to the gr::work() side.

I think the easier/simpler solution would be to handle this closer to the driver, also because  -- in case of using local timing -- the timestamp is closer to the actual trigger compared to when taking the time-stamp in the GR work function which may be invoked a couple of 1-10 milliseconds later. E.g.

 1. FPGA produces M_channel x N_samples 16 bit raw data arrays
 2. using the 16-bit data chunks -> detect trigger ( `... & F` != 0`) + local-time-stamp & shift input by 4 bits 
 3. copy the raw data to 2 x internal circular_buffers (1 for data + 1 for tags 
 4. in GR work(...) function: 
   * read from the internal circular buffers
   * add tag meta information if available on msg port
     * SW trigger -> no wait just pass the data, 
     * HW trigger wait/programmable time-out 
   * copy the shifted data into the GR output buffers + adding the corresponding tags
   * set the number of produced samples in the output buffers
   * return 'OK' and voila
   
This way we also wouldn't need additional threads for handling the message port i.e. I think it's OK to block the GR work function for up to ~20 ms to wait for msg input data. Normally the msg info should arrive anyway first before the detected trigger.

@frankosterfeld What do you think?

Reply:
If it's ok to block the work() function, that would indeed simplify the implementation a lot, and one could move the reading/tag detection to a separate thread more easily, if needed. Just to to be sure, with "wait" you mean std::this_thread::sleep_for?

"FPGA produces M_channel x N_samples 16 bit raw data arrays" - note that LimeSuite (at least the Lms_RecvStream function that I've found) returns 2x16bit I/Q samples, not 1x16bit raw samples, I wonder if that's the reason for the confusion around the trigger bits?

"local-time-stamp" - In the absence of configured trigger bits, the logic is "tag the first sample with the data from the last timing message received" If there are no timing messages, no tags are assigned (same as for PicoScope). Should we rely on timestamps from the local machine again?

Reply:
"If it's ok to block the work() function, that would indeed simplify the implementation a lot" - that can't work like this I think, as simply sleep()ing would also block the timing message delivery. So we have to give the control back to the scheduler.